### PR TITLE
Removed Knative Buffering

### PR DESCRIPTION
**Pull request by juliababkina**
solution for https://github.com/pivotal/skenario/issues/61

### CODE REVIEW

File: pkg/model/cluster.go [hunk 1]
```diff
@@ -58,10 +59,11 @@ type clusterModel struct {
 	replicasActive      simulator.ThroughStock
 	replicasTerminating ReplicasTerminatingStock
 	replicasTerminated  simulator.SinkStock
-	requestsInBuffer    simulator.ThroughStock
-	requestsFailed      simulator.SinkStock
-	kubernetesClient    kubernetes.Interface
-	endpointsInformer   corev1informers.EndpointsInformer
+	//TODO think how to rename it
+	requestsInBuffer  simulator.ThroughStock
```
Comment (josephburnett) [inline pkg/model/cluster.go:63]: How about `requeststInRouting`?

File: pkg/model/cluster.go [hunk 2]
```diff
@@ -149,5 +151,11 @@ func NewCluster(env simulator.Environment, config ClusterConfig, replicasConfig
 
 	cm.replicasDesired = NewReplicasDesiredStock(env, desiredConf, cm.replicaSource, cm.replicasLaunching, cm.replicasActive, cm.replicasTerminating)
 
+	//start with 1 active replica
+	rs := cm.replicaSource.(*replicaSource)
+	for i := 0; i < config.InitialNumberOfReplicas; i++ {
+		replicasActive.Add(NewReplicaEntity(rs.env, rs.kubernetesClient, rs.endpointsInformer, rs.Next(), rs.maxReplicaRPS))
```
Comment (josephburnett) [inline pkg/model/cluster.go:157]: Created #68 to remove endpoints informer.  We can do it in a separate PR.  No action required, just FYI.

File: pkg/model/cluster_test.go [hunk 3]
```diff
@@ -172,9 +172,9 @@ func testCluster(t *testing.T, describe spec.G, it spec.S) {
 		})
 	})
 
-	describe("BufferStock()", func() {
-		it("returns the configured buffer stock", func() {
-			assert.Equal(t, rawSubject.requestsInBuffer, subject.BufferStock())
+	describe("RoutingStock()", func() {
+		it("returns the configured routing stock", func() {
+			assert.Equal(t, rawSubject.requestsInBuffer, subject.RoutingStock())
```
Comment (josephburnett) [inline pkg/model/cluster_test.go:177]: `requestsInRouting`

File: pkg/model/request_entity_test.go [hunk 4]
```diff
@@ -18,7 +18,7 @@ package model
 import (
 	"fmt"
 	"testing"
-	"time"
+	//"time"
```
Comment (josephburnett) [inline pkg/model/request_entity_test.go:21]: Delete.

File: pkg/model/request_entity.go [hunk 5]
```diff
@@ -34,11 +29,10 @@ type RequestEntity interface {
...
 type requestEntity struct {
-	env         simulator.Environment
-	number      int
-	bufferStock RequestsBufferedStock
-	nextBackoff time.Duration
-	attempts    int
+	env          simulator.Environment
+	number       int
+	routingStock RequestsRoutingStock
+	attempts     int
```
Comment (josephburnett) [inline pkg/model/request_entity.go:35]: We don't need `attempts` anymore.  The routing stock will send to an active replica or fail the request.  Just once.

File: pkg/model/request_entity_test.go [hunk 6]
```diff
@@ -49,13 +49,13 @@ func testRequestEntity(t *testing.T, describe spec.G, it spec.S) {
...
-		it("sets the buffer stock", func() {
-			assert.Equal(t, bufferStock, rawSubject.bufferStock)
+		it("sets the routing stock", func() {
+			assert.Equal(t, routingStock, rawSubject.routingStock)
 		})
 
-		it("sets the retry backoff duration to 100ms", func() {
-			assert.Equal(t, 100*time.Millisecond, rawSubject.nextBackoff)
-		})
+		//it("sets the retry backoff duration to 100ms", func() {
```
Comment (josephburnett) [inline pkg/model/request_entity_test.go:56]: Delete.

File: pkg/model/request_entity_test.go [hunk 7]
```diff
@@ -79,54 +79,54 @@ func testRequestEntity(t *testing.T, describe spec.G, it spec.S) {
...
-				assert.False(t, outOfAttempts)
-			})
-
-			it("returns true to indicate no more retries are possible", func() {
-				_, outOfAttempts = subject.NextBackoff()
-				assert.True(t, outOfAttempts)
-			})
-		})
-	})
+	//describe("NextBackoff()", func() {
```
Comment (josephburnett) [inline pkg/model/request_entity_test.go:82]: If these tests don't apply anymore, just delete them.

File: pkg/model/cluster.go [hunk 8]
```diff
@@ -30,9 +30,10 @@ import (
 )
 
 type ClusterConfig struct {
-	LaunchDelay      time.Duration
-	TerminateDelay   time.Duration
-	NumberOfRequests uint
+	LaunchDelay             time.Duration
+	TerminateDelay          time.Duration
+	NumberOfRequests        uint
+	InitialNumberOfReplicas int
```
Comment (jchesterpivotal) [inline pkg/model/cluster.go:36]: Suggest `uint` to catch unexpected negative values.

File: pkg/model/trafficpatterns/ramp.go [hunk 9]
```diff
@@ -65,7 +66,7 @@ func (r *ramp) Generate() {
 	}
 }
 
-func NewRamp(env simulator.Environment, source model.TrafficSource, buffer model.RequestsBufferedStock, config RampConfig) Pattern {
+func NewRamp(env simulator.Environment, source model.TrafficSource, buffer model.RequestsRoutingStock, config RampConfig) Pattern {
```
Comment (jchesterpivotal) [inline pkg/model/trafficpatterns/ramp.go:69]: Rename `buffer`.

File: pkg/model/trafficpatterns/sinusoidal.go [hunk 10]
```diff
@@ -61,7 +61,7 @@ func (s *sinusoidal) Generate() {
 	}
 }
 
-func NewSinusoidal(env simulator.Environment, source model.TrafficSource, buffer model.RequestsBufferedStock, config SinusoidalConfig) Pattern {
+func NewSinusoidal(env simulator.Environment, source model.TrafficSource, buffer model.RequestsRoutingStock, config SinusoidalConfig) Pattern {
```
Comment (jchesterpivotal) [inline pkg/model/trafficpatterns/sinusoidal.go:64]: Rename `buffer`.

File: pkg/model/cluster_test.go [hunk 11]
```diff
@@ -138,30 +138,30 @@ func testCluster(t *testing.T, describe spec.G, it spec.S) {
 			rawSubject.replicasActive.Add(secondReplica)
 
 			subject.RecordToAutoscaler(autoscalerFake, &theTime)
-			bufferRecorded = autoscalerFake.recorded[0]
+			routingStockRecorded = autoscalerFake.recorded[0]
 		})
 
-		// TODO immediately record arrivals at buffer
+		// TODO immediately record arrivals at rotingStock
```
Comment (jchesterpivotal) [inline pkg/model/cluster_test.go:144]: Typo here: `rotingStock` -> `routingStock`

File: pkg/model/cluster.go (hunk 1 above)
Comment (juliababkina) [reply pkg/model/cluster.go]: Renamed it in https://github.com/pivotal/skenario/pull/67/commits/dfd4d60851eea3b87cc9d7e4dc950b95bc92c495 commit

File: pkg/model/cluster.go (hunk 2 above)
Comment (juliababkina) [reply pkg/model/cluster.go:155]: Ok

File: pkg/model/cluster_test.go (hunk 3 above)
Comment (juliababkina) [reply pkg/model/cluster_test.go]: Renamed it in https://github.com/pivotal/skenario/pull/67/commits/dfd4d60851eea3b87cc9d7e4dc950b95bc92c495 commit

File: pkg/model/request_entity_test.go (hunk 4 above)
Comment (juliababkina) [reply pkg/model/request_entity_test.go]: Deleted it in https://github.com/pivotal/skenario/pull/67/commits/dfd4d60851eea3b87cc9d7e4dc950b95bc92c495 commit

File: pkg/model/request_entity.go (hunk 5 above)
Comment (juliababkina) [reply pkg/model/request_entity.go]: Deleted attempts field in https://github.com/pivotal/skenario/pull/67/commits/dfd4d60851eea3b87cc9d7e4dc950b95bc92c495 commit

File: pkg/model/request_entity_test.go (hunk 7 above)
Comment (juliababkina) [reply pkg/model/request_entity_test.go]: Deleted them in https://github.com/pivotal/skenario/pull/67/commits/dfd4d60851eea3b87cc9d7e4dc950b95bc92c495 commit.

File: pkg/model/cluster.go (hunk 8 above)
Comment (juliababkina) [reply pkg/model/cluster.go]: Did it in https://github.com/pivotal/skenario/pull/67/commits/dfd4d60851eea3b87cc9d7e4dc950b95bc92c495 commit.

File: pkg/model/trafficpatterns/ramp.go (hunk 9 above)
Comment (juliababkina) [reply pkg/model/trafficpatterns/ramp.go]: Renamed to routingStock in https://github.com/pivotal/skenario/pull/67/commits/dfd4d60851eea3b87cc9d7e4dc950b95bc92c495 commit.

File: pkg/model/trafficpatterns/sinusoidal.go (hunk 10 above)
Comment (juliababkina) [reply pkg/model/trafficpatterns/sinusoidal.go]: Renamed to routingStock in https://github.com/pivotal/skenario/pull/67/commits/dfd4d60851eea3b87cc9d7e4dc950b95bc92c495 commit.

File: pkg/model/request_entity_test.go (hunk 6 above)
Comment (juliababkina) [reply pkg/model/request_entity_test.go]: Deleted it in https://github.com/pivotal/skenario/pull/67/commits/b3c3d8657d2149597dd78d8f0701746ca53862f2 commit.

File: pkg/model/cluster_test.go (hunk 11 above)
Comment (juliababkina) [reply pkg/model/cluster_test.go]: Fixed in https://github.com/pivotal/skenario/pull/67/commits/b3c3d8657d2149597dd78d8f0701746ca53862f2 commit.

**Comment by josephburnett** [comment]
@juliababkina looks like the units aren't passing on this change yet.  They need to be fixed before I will merge this PR.

**Comment by josephburnett** [comment]
Looks like there is one more unit test failure: https://gist.githubusercontent.com/josephburnett/13f39b9bd1d3ca7c8104e79b3e21f533/raw/09104608232311b434200dbecae1a2f07952f6b1/gistfile1.txt
